File: Java/StreamView/src/model/IStreamModel.java
package model;

import java.io.File;
import java.io.IOException;

/*******************************************************************************
 * This model represents a file stream that is read and presented in blocks of
 * bytes.
 ******************************************************************************/
public interface IStreamModel
{
    /***************************************************************************
     * Returns the block size of this model.
     * 
     * @return
     **************************************************************************/
    int getBlockSize();

    /***************************************************************************
     * Returns the number of blocks in this model.
     * 
     * @return
     **************************************************************************/
    int getBlockCount();

    /***************************************************************************
     * Returns the current block index.
     * 
     * @return
     **************************************************************************/
    int getBlockIndex();

    /***************************************************************************
     * Returns the current block as an array of bytes.
     * 
     * @return
     **************************************************************************/
    byte[] getBlockBytes();

    /***************************************************************************
     * Returns the total size of this model.
     * 
     * @return
     **************************************************************************/
    long getTotalSize();

    /***************************************************************************
     * Reads the next block of data from the model.
     * 
     * @throws IOException
     **************************************************************************/
    void readNextBlock() throws IOException;

    /***************************************************************************
     * Reads the previous block of data from the model.
     * 
     * @throws IOException
     **************************************************************************/
    void readPrevBlock() throws IOException;

    /***************************************************************************
     * Reads the specified block of data from the model.
     * 
     * @param index
     * @throws IOException
     **************************************************************************/
    void readBlock( int index ) throws IOException;

    /***************************************************************************
     * Sets the input file to use as the source of data for this model.
     * 
     * @param in
     * @throws Exception
     **************************************************************************/
    void setInputFile( File in ) throws Exception;

    /***************************************************************************
     * Sets the input file and block size to use as the source of data for this
     * model.
     * 
     * @param in
     * @param blockSize
     * @throws Exception
     **************************************************************************/
    void setInputFile( File in, int blockSize ) throws Exception;
}

File: C#/Chex/Chex/Data/IStreamModel.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Chex.Data
{
    interface IStreamModel
    {
        #region EventHandlers

        event EventHandler BlockChanged;
        
        #endregion

        #region Properties

        /// <summary>
        /// The current stream.
        /// </summary>
        Stream CurrentStream { get; }

        /// <summary>
        /// The block size.
        /// </summary>
        uint BlockSize { get; }

        /// <summary>
        /// The number of blocks in the current stream.
        /// </summary>
        uint BlockCount { get; }

        /// <summary>
        /// The index of the current block.
        /// </summary>
        uint BlockIndex { get; }

        /// <summary>
        /// The current block of bytes.
        /// </summary>
        IList<Byte> BlockBytes { get; }

        /// <summary>
        /// Previous block available?
        /// </summary>
        bool HasPrevBlock { get; }

        /// <summary>
        /// Next block available?
        /// </summary>
        bool HasNextBlock { get; }

        #endregion

        #region Methods

        /// <summary>
        /// Reads the next block of data from the underlying stream.
        /// Does nothing if the next block does not exist.
        /// </summary>
        void readNextBlock();

        /// <summary>
        /// Reads the previous block of data from the underlying stream.
        /// Does nothing if the previous block does not exist.
        /// </summary>
        void readPrevBlock();

        /// <summary>
        /// Reads the specified block of data from the stream.
        /// Does nothing if the block does not exist.
        /// </summary>
        /// <param name="index"></param>
        void readBlock( uint index );

        /// <summary>
        /// Sets the underlying stream from which data is read.
        /// </summary>
        /// <param name="stream"></param>
        void setInputStream( Stream stream );

        /// <summary>
        /// Writes the underlying stream.
        /// </summary>
        void writeStream();

        /// <summary>
        /// Write the current stream to <code>newStream</code>, then changes
        /// the current stream to <code>newStream</code>.
        /// </summary>
        /// <param name="stream"></param>
        void writeStream( Stream newStream );

        /// <summary>
        /// Returns true if the block index is valid.
        /// </summary>
        /// <returns></returns>
        bool isValidBlock( uint block );

        /// <summary>
        /// Finds the next byte that matches the search key.
        /// </summary>
        /// <param name="offset"></param>
        /// <param name="searchKey"></param>
        /// <returns>The index of the next byte after <code>offset</code> that
        /// matches <code>searchKey</code>, or -1 if no match is found.</returns>
        int find( uint offset, byte searchKey );

        /// <summary>
        /// Finds the previous byte that matches the search key.
        /// </summary>
        /// <param name="offset"></param>
        /// <param name="searchKey"></param>
        /// <returns>The index of the next byte before <code>offset</code> that
        /// matches <code>searchKey</code>, or -1 if no match is found.</returns>
        int rfind( uint offset, byte searchKey );

        /// <summary>
        /// Loads the block that contains <code>offset</code>.
        /// </summary>
        /// <param name="offset"></param>
        /// <returns>The index of the desired offset in the current block,
        /// or -1 if <code>offset</code> is invalid.</returns>
        int gotoOffset( uint offset );

        #endregion
    }
}
